`haskellPackages.aeson_2_2_0_0`: fix build

peterbecich opened the pull request:
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/254183

```
nix build .#haskellPackages.aeson_2_2_0_0
```
succeeds.



## Things done

Addition to `configuration-nix.nix` for this package.

Ran
```
./maintainers/scripts/haskell/regenerate-hackage-packages.sh
```

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cdepillabout commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1229,6 +1229,26 @@ self: super: {
   # Tests require newer package version.
   aeson_1_5_6_0 = dontCheck (doJailbreak super.aeson_1_5_6_0);
 
+  aeson_2_2_0_0 = super.aeson_2_2_0_0.override {
+    th-abstraction = self.th-abstraction_0_5_0_0;
+  };
+
+  generic-deriving = super.generic-deriving.override {
+    th-abstraction = self.th-abstraction_0_5_0_0;
+  };
+
+  th-lift = super.th-lift.override {
+    th-abstraction = self.th-abstraction_0_5_0_0;
+  };
+
+  th-expand-syns = super.th-expand-syns.override {
+    th-abstraction = self.th-abstraction_0_5_0_0;
+  };
+
+  bifunctors = super.bifunctors.override {
+    th-abstraction = self.th-abstraction_0_5_0_0;
+  };
```
Since things like `generic-deriving`, `th-lift`, etc already compile using the current version of `th-abstraction`, we shouldn't override the `th-abstraction` version in all of these packages (since that will make them incompatible with any packages _not_ overriding the `th-abstraction` version).

If I understand correctly, your goal with this is to get `spago` compiling correctly, right?

In that case, it would probably be better to throw `th-abstraction` into the `.overrideScope` for `spago` in `pkgs/development/haskell-modules/configuration-nix.nix`.  That way, these changes would only affect spago (and not the rest of the package set).

Also, can you make sure your commit message and PR title matches the format specified in https://github.com/NixOS/nixpkgs/blob/9108a20782535741433c304f6a4376cb8b364b89/CONTRIBUTING.md#commit-conventions?

peterbecich replied:
Thanks, yes, I am attempting to fix Spago: https://github.com/NixOS/nixpkgs/issues/253198

However, I am also interested in specifically fixing the `aeson_2_2_0_0` package, regardless of Spago.

cdepillabout replied:
> However, I am also interested in specifically fixing the aeson_2_2_0_0 package, regardless of Spago.

In that case, maybe using an `.overrideScope` on `aeson_2_2_0_0` for `th-abstraction_0_5_0_0` would be the simplest solution.

peterbecich committed:
haskellPackages.aeson_2_2_0_0: fix build

aeson 2.2.0.0 requires th-abstraction >= 0.5 & < 0.6

Ran
```
./maintainers/scripts/haskell/regenerate-hackage-packages.sh
```

Now this is successful:
```
nix build .#haskellPackages.aeson_2_2_0_0
```

Closes: https://github.com/NixOS/nixpkgs/issues/254183

cdepillabout commented:
Thanks, I fixed this up and merged it in as 7fc032e5951b into https://github.com/NixOS/nixpkgs/pull/249708.
